File: StateOfCharge/getBatteryParameters.m
function [R_in, R_CT, C_CT] = getBatteryParameters(file,CRate)

% Load data from file (.CSV) formatted as below
% {'TimeStamp', 'TimeDelta', 'InputCurrent', 'OutputVoltage'}
data = readtable(file);

idx_Time = 1;
idx_DeltaT = 2;
idx_InCurr = 3;
idx_VoltOut = 4;

% Time for the 1st pulse to start (seconds)
T_pulse_start = 50;
% duration of pulse (seconds) for 10% SOC
T_pulse_end = 360/CRate + T_pulse_start;

I_Rin_0 = 0;
V_Rin_0 = 0;
V_R_CT_0 = 0;
I_R_CT = 0;
T_Rin = 0;
T_RCT = 0;
computeStage = 0;

n = height(data);
prev_Volt = 0;

for i = 2:n
    
    if data{i, idx_Time} > T_pulse_start
        
        deltaV = data{i, idx_VoltOut} - prev_Volt;
        
        if abs(deltaV) > 0.05 && computeStage == 0
            
            I_Rin_0 = data{i, idx_InCurr};
            V_Rin_0 = data{i-1, idx_VoltOut};
            T_Rin = data{i-1, idx_Time};
            computeStage = 1;
            continue
            
        end
        
        if computeStage == 1
            
            if data{i, idx_Time} - T_Rin > 0.015
                
                R_in = (V_Rin_0 - data{i, idx_VoltOut})/I_Rin_0;
                T_RCT = data{i, idx_Time} + 1 ;
                computeStage = 2;
                continue
            
            end
        end
        
        if computeStage == 2 && data{i, idx_Time} > T_RCT
            
             V_R_CT_0 = data{i, idx_VoltOut};
             I_R_CT = data{i, idx_InCurr};
             computeStage = 3;
             continue;
            
        end
        
        if computeStage == 3 && deltaV > 0 && data{i, idx_Time} > T_pulse_end
            
            R_CT = (V_R_CT_0 - data{i-1, idx_VoltOut})/I_R_CT - R_in;
            T_RCT = data{i, idx_Time};
            computeStage = 4;
            continue
            
        end

        if computeStage == 4 && deltaV ~= 0 && deltaV < 0.00009
    
            C_CT = (data{i, idx_Time} - T_RCT)/R_CT;
            break

        end
        
    end
    
    prev_Volt = data{i, idx_VoltOut};
    
end

end

File: _BatteryTest_BlueSky/get_battery_parameters.py
import pandas as pd
import time, os


def get_battery_parameters(in_file_path, out_dir, CRate):
    # Load data from in_file_path (.CSV) formatted as below
    # ['TimeStamp', 'TimeDelta', 'InputCurrent', 'OutputVoltage']
    data = pd.read_csv(in_file_path)

    idx_Time = 0
    idx_DeltaT = 1
    idx_InCurr = 2
    idx_VoltOut = 3

    # Time for the 1st pulse to start (seconds)
    T_pulse_start = 50
    # duration of pulse (seconds) for 10% SOC
    T_pulse_end = 360 / CRate + T_pulse_start

    I_Rin_0 = 0
    V_Rin_0 = 0
    V_R_CT_0 = 0
    I_R_CT = 0
    T_Rin = 0
    T_RCT = 0
    compute_stage = 0

    n = len(data)
    prev_Volt = 0

    for i in range(1, n):
        if data.iloc[i, idx_Time] > T_pulse_start:
            deltaV = data.iloc[i, idx_VoltOut] - prev_Volt

            if abs(deltaV) > 0.05 and compute_stage == 0:
                I_Rin_0 = data.iloc[i, idx_InCurr]
                V_Rin_0 = data.iloc[i-1, idx_VoltOut]
                T_Rin = data.iloc[i-1, idx_Time]
                compute_stage = 1
                continue

            if compute_stage == 1 and data.iloc[i, idx_Time] - T_Rin > 0.015:
                R_in = (V_Rin_0 - data.iloc[i, idx_VoltOut]) / I_Rin_0
                T_RCT = data.iloc[i, idx_Time] + 1
                compute_stage = 2
                continue

            if compute_stage == 2 and data.iloc[i, idx_Time] > T_RCT:
                V_R_CT_0 = data.iloc[i, idx_VoltOut]
                I_R_CT = data.iloc[i, idx_InCurr]
                compute_stage = 3
                continue

            if compute_stage == 3 and deltaV > 0 and data.iloc[i, idx_Time] > T_pulse_end:
                R_CT = (V_R_CT_0 - data.iloc[i-1, idx_VoltOut]) / I_R_CT - R_in
                T_RCT = data.iloc[i, idx_Time]
                compute_stage = 4
                continue

            if compute_stage == 4 and deltaV != 0 and deltaV < 0.00009:
                C_CT = (data.iloc[i, idx_Time] - T_RCT) / R_CT
                current_time = time.localtime()
                out_file_name = 'battery_parameters' + str(current_time.tm_hour) + str(current_time.tm_min) + '.csv'
                out_table = pd.DataFrame({'R_in': [R_in], 'R_CT': [R_CT], 'C_CT': [C_CT]})
                if not os.path.exists(out_dir):
                    os.makedirs(out_dir)
                out_table.to_csv(os.path.join(out_dir, out_file_name))
                return out_table

        prev_Volt = data.iloc[i, idx_VoltOut]